File: android/app/src/main/java/online/db1k/safering/android/service/HelpSignaler.kt
package online.db1k.safering.android.service

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri

enum class HelpReason(val shortLabel: String) {
    MONEY("someone asking for money"),
    HELP("tapped Help"),
    AFTER_CALL("a call just ended and they want a check-in"),
    PASTE_SCAM("they pasted a message that looks like a scam"),
    VERIFY("they want you to verify a call or message");
}

class HelpSignaler(
    private val context: Context,
    private val household: HouseholdStore
) {
    fun draftBody(reason: HelpReason): String {
        val name = household.ownerDisplayName.ifBlank { "Your person" }
        val trusted = household.trustedContactName.ifBlank { "there" }
        return """
            SafeRing alert for $trusted:

            $name — ${reason.shortLabel}.

            Call them on their saved number. Do not call back an unknown number. Do not send money or gift cards.

            This is a redacted family alert. No caller ID or message content is included.
        """.trimIndent()
    }

    fun send(reason: HelpReason) {
        when (household.preferredChannel) {
            SignalChannel.SMS -> openSms(reason)
            SignalChannel.WHATSAPP -> openWhatsApp(reason)
            SignalChannel.SIGNAL -> openSignal()
            SignalChannel.PHONE -> callSaved()
        }
        household.recordHelpSent()
    }

    fun openSms(reason: HelpReason) {
        val number = household.e164Number
        val intent = Intent(Intent.ACTION_SENDTO).apply {
            data = Uri.parse("smsto:$number")
            putExtra("sms_body", draftBody(reason))
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        context.startActivity(intent)
    }

    fun callSaved() {
        val number = household.e164Number
        val intent = Intent(Intent.ACTION_DIAL).apply {
            data = Uri.parse("tel:$number")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        context.startActivity(intent)
    }

    fun openSignal() {
        val number = household.e164Number
        val uri = Uri.parse("sgnl://signal.me/#p/$number")
        val intent = Intent(Intent.ACTION_VIEW, uri).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        if (canHandle(intent)) {
            context.startActivity(intent)
            return
        }
        val web = Intent(Intent.ACTION_VIEW, Uri.parse("https://signal.me/#p/$number"))
            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context.startActivity(web)
    }

    fun openWhatsApp(reason: HelpReason) {
        val digits = household.e164Number.filter { it.isDigit() }
        val text = Uri.encode(draftBody(reason))
        val uri = Uri.parse("https://wa.me/$digits?text=$text")
        val intent = Intent(Intent.ACTION_VIEW, uri).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context.startActivity(intent)
    }

    fun isInstalled(channel: SignalChannel): Boolean {
        val pkg = when (channel) {
            SignalChannel.SIGNAL -> "org.thoughtcrime.securesms"
            SignalChannel.WHATSAPP -> "com.whatsapp"
            else -> return true
        }
        return try {
            context.packageManager.getPackageInfo(pkg, 0)
            true
        } catch (_: PackageManager.NameNotFoundException) {
            false
        }
    }

    private fun canHandle(intent: Intent): Boolean {
        return intent.resolveActivity(context.packageManager) != null
    }
}

File: ios/SafeRing/Service/HelpSignaler.swift
import Foundation
import UIKit
import MessageUI

/// Off-call signaling to the trusted contact.
///
/// There is no Signal send API and no RCS API a third-party iOS app can call.
/// Order of usefulness on iPhone:
/// 1. SMS / iMessage via MessageUI (prefilled, works with any number)
/// 2. WhatsApp URL (can prefill text if installed)
/// 3. FaceTime Audio (encrypted live verify, not a text alert)
/// 4. Signal URL (opens a chat, cannot prefill)
enum HelpReason: String {
    case money
    case help
    case afterCall
    case checkInTimeout
    case pasteScam
    case verify

    var shortLabel: String {
        switch self {
        case .money: return "someone asking for money"
        case .help: return "tapped Help"
        case .afterCall: return "a call just ended and they want a check-in"
        case .checkInTimeout: return "a call ended and they did not confirm it was okay"
        case .pasteScam: return "they pasted a message that looks like a scam"
        case .verify: return "they want you to verify a call or message"
        }
    }
}

struct HelpDraft: Identifiable, Equatable {
    let id = UUID()
    let recipients: [String]
    let body: String
    let reason: HelpReason
}

enum HelpSignalError: LocalizedError {
    case notConfigured
    case messagingUnavailable
    case cannotOpen(String)

    var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "Add a trusted person first."
        case .messagingUnavailable:
            return "This device cannot send texts."
        case .cannotOpen(let app):
            return "Could not open \(app). Is it installed?"
        }
    }
}

@MainActor
final class HelpSignaler: ObservableObject {

    static let shared = HelpSignaler()

    @Published var draft: HelpDraft?
    @Published var lastError: String?

    private let household: HouseholdStore

    init(household: HouseholdStore = .shared) {
        self.household = household
    }

    func makeDraft(reason: HelpReason) throws -> HelpDraft {
        guard household.isConfigured else { throw HelpSignalError.notConfigured }
        let name = household.ownerDisplayName
        let trusted = household.trustedContactName.isEmpty ? "there" : household.trustedContactName
        let body = """
        SafeRing alert for \(trusted):

        \(name) — \(reason.shortLabel).

        Call them on their saved number. Do not call back an unknown number. Do not send money or gift cards.

        This is a redacted family alert. No caller ID or message content is included.
        """
        return HelpDraft(
            recipients: [household.e164Number],
            body: body,
            reason: reason
        )
    }

    func prepareSMS(reason: HelpReason) throws {
        draft = try makeDraft(reason: reason)
        guard MFMessageComposeViewController.canSendText() else {
            throw HelpSignalError.messagingUnavailable
        }
    }

    func openPreferred(reason: HelpReason) throws {
        switch household.preferredChannel {
        case .sms:
            try prepareSMS(reason: reason)
        case .whatsapp:
            try openWhatsApp(reason: reason)
        case .signal:
            try openSignal()
        case .facetime:
            // FaceTime is Apple-only. Help alerts use SMS so Android numbers work.
            try prepareSMS(reason: reason)
        }
    }

    func openFaceTimeAudio() throws {
        guard household.isConfigured else { throw HelpSignalError.notConfigured }
        let number = household.e164Number
        let candidates = [
            URL(string: "facetime-audio://\(number)"),
            URL(string: "facetime://\(number)"),
        ].compactMap { $0 }
        for url in candidates where UIApplication.shared.canOpenURL(url) {
            UIApplication.shared.open(url)
            household.recordHelpSent()
            return
        }
        try openPhone()
    }

    func openPhone() throws {
        guard household.isConfigured else { throw HelpSignalError.notConfigured }
        // Real cellular/PSTN call — reaches Android. Never FaceTime on this path.
        let raw = household.e164Number
        let allowed = Set("+0123456789")
        let digits = String(raw.filter { allowed.contains($0) })
        guard digits.contains(where: { $0.isNumber }) else {
            throw HelpSignalError.notConfigured
        }
        // tel:+E164 (not tel://) — reliable on modern iOS
        guard let url = URL(string: "tel:\(digits)") else {
            throw HelpSignalError.cannotOpen("Phone")
        }
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }

    func openSignal() throws {
        guard household.isConfigured else { throw HelpSignalError.notConfigured }
        let number = household.e164Number
        let candidates = [
            URL(string: "sgnl://signal.me/#p/\(number)"),
            URL(string: "https://signal.me/#p/\(number)"),
        ].compactMap { $0 }
        for url in candidates {
            if url.scheme == "https" || UIApplication.shared.canOpenURL(url) {
                UIApplication.shared.open(url)
                household.recordHelpSent()
                return
            }
        }
        throw HelpSignalError.cannotOpen("Signal")
    }

    func openWhatsApp(reason: HelpReason) throws {
        let draft = try makeDraft(reason: reason)
        let digits = household.e164Number.filter(\.isNumber)
        let encoded = draft.body.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        let candidates = [
            URL(string: "whatsapp://send?phone=\(digits)&text=\(encoded)"),
            URL(string: "https://wa.me/\(digits)?text=\(encoded)"),
        ].compactMap { $0 }
        for url in candidates {
            if url.scheme == "https" || UIApplication.shared.canOpenURL(url) {
                UIApplication.shared.open(url)
                household.recordHelpSent()
                return
            }
        }
        throw HelpSignalError.cannotOpen("WhatsApp")
    }

    func isAvailable(_ channel: HouseholdStore.SignalChannel) -> Bool {
        switch channel {
        case .sms:
            return MFMessageComposeViewController.canSendText()
        case .whatsapp:
            return UIApplication.shared.canOpenURL(URL(string: "whatsapp://send")!)
        case .signal:
            return UIApplication.shared.canOpenURL(URL(string: "sgnl://signal.me")!)
        case .facetime:
            return UIApplication.shared.canOpenURL(URL(string: "facetime-audio://")!)
                || UIApplication.shared.canOpenURL(URL(string: "facetime://")!)
        }
    }

    func markSent() {
        household.recordHelpSent()
        draft = nil
    }

    func cancelDraft() {
        draft = nil
    }
}
